### DISCUSSION TITLE
all questions regarding PR#13

### DISCUSSION THREAD

**Discussion by jkowalleck** [CycloneDX/cyclonedx-node-yarn#14] [Ideas]
all questions from https://github.com/CycloneDX/cyclonedx-node-yarn/pull/13 could go below

**Comment by jkowalleck** [comment]
Q: What are sensible defaults for the parameters?

**Reply by jkowalleck** [reply]
* `--spec-version` should default to `1.5`
* `--output-format` should default to `JSON` -- its just native to the JavaScript world, 
* `--output-file` should default to printing to STDOUT
* `--component-type` should default to `application`
* `--production` should default to `false`

**Comment by jkowalleck** [comment]
Q: Decision on license. Am I correct to assume the contained components can be used as part of a Apache-2.0 licensed project? Are you aware of any tool that takes a list of SPDX identifiers for dependencies (or SBOM) and answers which licenses could be chosen?

**Reply by jkowalleck** [reply]
In which context? In the context of the license for the distributional of the yarn plugin?  
I understand that a yarn plugin is just one JavaScript file - which would be an assembly of multiple components and some own code. Leading to a license file disclosing all used components and stating their licenses. At least this is what I would assume. But in case this is the question, let me know and i will ask the OWASP legals team for guidance.

**Reply by jkowalleck** [reply]
well, i think we need to considder two things here: 
1. what is the license of the code that is hosted in this repository? 
2. what is the license of the build result?

Answer to 1: Apache-2.0

Answer to 2: nothing to bother now. as long as all components' licenses are properly listed with the aggregated build result, we are good for now.

**Reply by AugustusKling** [reply]
I was thinking about license for this repository which is a non-issue as Apache-2.0 is fine.

The more difficult part is the packaging and license for the compiled bundled. Since they contain at least parts of the dependencies one needs to provide a component listing and license texts to confirm with the licenses of the dependencies. Also the license for the bundle must safisfy all dependency licenses. My current understanding is that including `./bundles/components-licenses.md` in the bundle is okay given the current dependencies but maybe linking to the file is good enough.

However if direct or transitive dependencies are change, this may no longer hold true. Manual verification is tedious and error-prone. I'd hope some tool exists that can be fed SPDX identifiers of the desired license and the licenses of all dependencies. It should then break the build in case anything is incompatible. Perhaps a build step calling https://github.com/jslicense/spdx-satisfies.js would work?

**Reply by jkowalleck** [reply]
I do understand the situation.  
Since this is legal territory, I'd stop thinking about this. Should lawyers guide us.

I will seek the legal support from OWASP for guidance.

**Reply by AugustusKling** [reply]
Thanks. I'm curious to learn what OSWAP is going to suggest regarding component listings + licenses of dependencies for bundles.

**Reply by jkowalleck** [reply]
After some modifications to the CycloneDX library, it was possible to help esbuild with the tree shaking, so no unintended components were bundled.
Then, it required some tweaks to the yarn plugin builder, so that it was possible to gather the actual bundled components and their respective licenses.

Then i worked to a state where only apache2 compatible licenses were bundled. Then I talked with my OWASP colleagues and we assessed the licenses and compliance.

final result: we are ready to publish a bundle/assembly.
And this is where we are now 🎉 finally.


see also: 
- https://github.com/CycloneDX/cyclonedx-node-yarn/issues/104
- https://github.com/CycloneDX/cyclonedx-node-yarn/pull/117
- https://github.com/CycloneDX/cyclonedx-node-yarn/blob/1.0-dev/tools/write-3rd-party-licenses.cjs
- https://github.com/yarnpkg/berry/issues/6211

**Comment by jkowalleck** [comment]
Q: Treatment of the root component in the SBOM. Should it be put in the metadata only or also within the components property?

**Reply by jkowalleck** [reply]
especially in `$.metadata.tools[]`

**Comment by jkowalleck** [comment]
Q: Many files in the root folder are basically empty templates.

**Reply by jkowalleck** [reply]
which in particular? what about those files?

**Reply by AugustusKling** [reply]
* .gitignore
* CODEOWNERS, linked https://github.com/orgs/CycloneDX/teams/javascript-maintainers does not exist.
* CONTRIBUTING.md
* HISTORY.md

Maybe the files should only be created from the templates when putting content in that is specific to this repository.

**Reply by jkowalleck** [reply]
> .gitignore

was boilerplate with a basic setup for this project.
feel free to modify to your needs.

> CODEOWNERS, linked https://github.com/orgs/CycloneDX/teams/javascript-maintainers does not exist.

it does exist, it is just no group you might have access to.

> CONTRIBUTING.md

it was prepared to be used, not a template. is just had some gaps, since there was no processes for some cases, since the was no code yet. 

feel free to fill those gaps

> HISTORY.md

it was prepared to be used, not a template. since there was no history, there was just no details. ;-)

**Comment by jkowalleck** [comment]
Q: Do you have sample projects with known expected output regarding listed components and licenses?

**Reply by jkowalleck** [reply]
nope, not for yarn. feel free to establish all the examples and test beds you need - for (unit/functional/integration) test or showcasing purposes

**Comment by jkowalleck** [comment]
Q: Fallback for component licenses.

**Reply by jkowalleck** [reply]
in the SBOM? none in particular, `component.licenses` is an optional list.

**Reply by AugustusKling** [reply]
Do you intent to have a dependency on something like https://github.com/CycloneDX/license-scanner in some distant future as part of this project?

**Reply by jkowalleck** [reply]
never say no. if there was a reason for it, a feasible way and a reasonable effort, why not.
but this would require them having a stable programmable interface, see https://github.com/CycloneDX/license-scanner/issues/9

**Comment by jkowalleck** [comment]
Q: How would one calculate the hash for https://cyclonedx.org/docs/1.5/json/#components_items_hashes for a folder representing a component?

**Reply by jkowalleck** [reply]
just don't for now. this would not be part of an minimal viable product. leave this feature for future development.

the hash or the external reference to the distributional instead, this one should be feasible, as it is in most yarn locks already.

**Reply by AugustusKling** [reply]
I don't know how the checksum from the yarn.lock is calculated nor how stable it is across lockfile versions. Reading bug reports like https://github.com/yarnpkg/berry/issues/6068 and https://github.com/yarnpkg/berry/issues/6105 suggests that the algorithm of the checksum is not defined.

**Reply by jkowalleck** [reply]
well, then leave it for now. can be added in the future, after the details are clear.

**Comment by jkowalleck** [comment]
Q: Which are the important/desired properties in the generated SBOM files besides those marked as required in the JSON schema?

**Reply by jkowalleck** [reply]
Whatever you feel relevant for a minimal viable product. :-) 
Future development may add additional features/information, so there is no need for a complete solution right away.

**Comment by jkowalleck** [comment]
Q: Is the reproducible results flag mentioned in https://github.com/CycloneDX/cyclonedx-node-yarn/discussions/8 important? Shouldn't this be a task for a SBOM consumer by masking out properties?

**Reply by jkowalleck** [reply]
the "reproducible" is more than masking out. see https://cyclonedx.github.io/cyclonedx-property-taxonomy/cdx.html
> Whether the CycloneDX document has been generated in a reproducible manner: if so, then time- or random-based values MUST be omitted, and elements order SHOULD be reproducible. Boolean value. MAY appear only once. SHOULD be used in `$.metadata.properties`.

Therefore, it is a core feature of this tool.

It indicates, that some elements were left out on purpose, and it indicates that all elements are on a fixed order.
Bringing things to order is easily achievable via the JS library's serializers' option `sortLists`. See an example: https://github.com/CycloneDX/cyclonedx-node-npm/blob/c7414ee0d032ba9d067001c2c27b42581b553c4b/src/cli.ts#L276-L279

----


in addition, having this feature is a hue benefit when it comes to testing:
you could prepare certain test beds and run this very tool over it and the resulting SBOM will always look the same, making it an ideal integration test element.

**Reply by AugustusKling** [reply]
CLI parameter `--reproducible` is now there. The `CDX.Models.Metadata` does not allow to specify properties, though. How do I set the propery using your API?

**Reply by jkowalleck** [reply]
> The `CDX.Models.Metadata` does not allow to specify properties, though. How do I set the propery using your API?

what do you need a property for? what do you mean?

**Reply by AugustusKling** [reply]
To quote your quote :wink:

> ... SHOULD be used in $.metadata.properties.

Since I cannot find generic properties in `CDX.Models.Metadata`, I wonder how this `cdx:reproducible` could be put there. Say we have `const bom = new CDX.Models.Bom();` how would one set the flag to express that the SBOM was created with `--reproducible`. Can you give a code example?

**Reply by jkowalleck** [reply]
something like  (untested pseudocode)
```js
bom.metadata.properties.push(new CDX.Models.Property('cdx:reproducible', 'true'))
```


PS: atleast if the properties would be implemented.
seams like nobodu had a use case for them before, so there is no implementation for it, yet -- https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/src/models/metadata.ts

**Reply by jkowalleck** [reply]
if you find a feature being missing from the library, feel free to create a ticket for it. :-) 

the library is a living implementation. each feature is added when needed. This enables the library to be published/available as soon as possible. New features are added on demand. ;-)

**Reply by AugustusKling** [reply]
You code example is what I had attempted but it didn't work as the model was incomplete. See https://github.com/CycloneDX/cyclonedx-javascript-library/issues/1019 for a potential fix.

**Reply by AugustusKling** [reply]
@jkowalleck, Property was added with b8ff4f7be7be1bb9627477e9a1ff544f3659968b

**Comment by jkowalleck** [comment]
Q: I've only used this plugin with Yarn 4.1.0. It may or may not work with other Yarn versions.

**Reply by jkowalleck** [reply]
One could add some automated tests to check. what i have un mind is a setup as follows: 

use github workflows as a automation for:
1. build
   1. setup the project in this repo
   1. build the project
   1. artifact the build result
2. tests: have a test matrix with all relevant versions of `yarn` and all relevant versions of `node`
   1. setup node in the version desired by the matrix
   2. setup yarn in the version desired by the matrix 
   4. fetch the artifact from the "build" step and use it for the following tests
   5. setup the project in this repo (skip build!)
   6. run all unit/functional/integration tests in an automated manner
2. dog-fooding: have a test matrix with all relevant versions of `yarn` and all relevant versions of `node`
   1. setup node in the version desired by the matrix
   2. setup yarn in the version desired by the matrix 
   3. fetch the artifact from the "build" step and install it 
   4. setup the project in this repo  (skip build!)
   7. run the yarn plugin over this very setup, and print the result to screen


similar is done here: https://github.com/CycloneDX/cyclonedx-node-npm/blob/main/.github/workflows/nodejs.yml
